src/main/java/com/celements/navigation/cmd/GetNotMappedMenuItemsForParentCommand.java

```diff
@@ -35,31 +37,64 @@
 import com.xpn.xwiki.XWikiException;
 import com.xpn.xwiki.doc.XWikiDocument;
 import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.web.Utils;
 
 public class GetNotMappedMenuItemsForParentCommand {
 
-  private static Log LOGGER = LogFactory.getFactory().getInstance(
+  private static Logger LOGGER = LoggerFactory.getLogger(
       GetNotMappedMenuItemsForParentCommand.class);
 
-  private HashMap<String, Map<String, List<TreeNode>>> menuItems;
+  private final Map<String, Map<String, List<TreeNode>>> menuItems;
   
   private int queryCount = 0;
 
+  private XWikiContext getContext() {
+    return (XWikiContext) Utils.getComponent(Execution.class).getContext(
+        ).getProperty("xwikicontext");
+  }
+
   public GetNotMappedMenuItemsForParentCommand() {
-    menuItems = new HashMap<String, Map<String, List<TreeNode>>>();
+    /* ConcurrentHashMap is needed to prevent:
+     * 1) ConcurrentModificationException if Thread A iterates over menuItems and
+     *    Thread B is changing it.
+     * 2) Thread B changes the HashMap and Thread B does not synchronize on read
+     *    which may lead that it reads an old version from its memory cache.
+     *    Details see JSR133
+     */
+    menuItems = new ConcurrentHashMap<String, Map<String, List<TreeNode>>>();
   }
 
-  HashMap<String, Map<String, List<TreeNode>>> getMenuItemsCache() {
+  Map<String, Map<String, List<TreeNode>>> getMenuItemsCache() {
     return menuItems;
   }
 
+  /**
+   * @deprecated instead use getTreeNodesForParentKey(String)
+   */
+  @Deprecated
   public List<TreeNode> getTreeNodesForParentKey(String searchParentKey,
       XWikiContext context) {
-    String cacheKey = getCacheKey(searchParentKey, context);
-    LOGGER.trace("getNotMappedMenuItemsFromDatabase: for cacheKey [" + cacheKey + "].");
-    if (!menuItems.containsKey(cacheKey)) {
+    return getTreeNodesForParentKey(searchParentKey);
+  }
+
+  public List<TreeNode> getTreeNodesForParentKey(String searchParentKey) {
+    String wikiName = getWikiName(searchParentKey);
+    LOGGER.trace("getNotMappedMenuItemsFromDatabase: for cacheKey [{}].", wikiName);
+    if (!menuItems.containsKey(wikiName)) {
+      loadMenuForWiki(wikiName);
+    }
+    if (menuItems.containsKey(wikiName) && (menuItems.get(wikiName) != null)
+        && (menuItems.get(wikiName).get(searchParentKey) != null)){
+      return menuItems.get(wikiName).get(searchParentKey);
+    }
```

Comment:
This if block is not thread safe. If `loadMenuForWiki` or `flushMenuItemCache` are called by other threads in between these statements, data inconsistency or even NPEs could occur.

I would suggest the following changes:
```
    Map<String, List<TreeNode>> wikiMenu = menuItems.get(wikiName);
    if (wikiMenu == null) {
      wikiMenu = loadMenuForWiki(wikiName);
    }
    if (wikiMenu.containsKey(searchParentKey)) {
      return wikiMenu.get(searchParentKey);
    }
```